Do not send two END_SERVICE requests during streaming switching between apps

LitvinenkoIra opened the pull request:
Fixes #3328 

This PR is **[not ready]** for review.

### Risk
This PR makes **[no]** API changes.

### Testing Plan
[ 3139_04_Stop_streaming_forcibly_app_does_not_send_EndServiceAck](https://github.com/smartdevicelink/sdl_atf_test_scripts/blob/develop/test_scripts/Defects/6_1/3139_04_Stop_streaming_forcibly_app_does_not_send_EndServiceAck.lua) 

### Summary
Check if EndService has been sent before sending it during unregistration

### Tasks Remaining:
- [x] Check regression
- [x] Check unit tests

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

AByzhynar commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -3215,7 +3215,10 @@ void ApplicationManagerImpl::UnregisterApplication(
+  if (navi_app_to_stop_.end() ==
+      std::find(navi_app_to_stop_.begin(), navi_app_to_stop_.end(), app_id)) {
+    EndNaviServices(app_id);
+  }
```
@LitvinenkoIra Could you be so kind and explain the logic here, please

LitvinenkoIra replied:
@AByzhynar In `EndNaviServices` function we add the `app_id` of application that should complete the services to the `navi_app_to_stop_` collection.
 So here we just need to check - if `app_id` already presents in  `navi_app_to_stop_`  collection -  we don’t need to call EndNaviServices, because we already called it once

LitvinenkoIra commented:
@theresalech this PR is ready for review. Thank you

jacobkeeler commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -3887,6 +3890,8 @@ void ApplicationManagerImpl::CloseNaviApp() {
     UnregisterApplication(app_id, ABORTED);
   }
+
+  navi_app_to_stop_.pop_front();
```
Seems this is more likely to cause a race condition, perhaps it should be lock-protected?

jacobkeeler commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -3215,7 +3215,10 @@ void ApplicationManagerImpl::UnregisterApplication(
 
-  EndNaviServices(app_id);
+  if (navi_app_to_stop_.end() ==
+      std::find(navi_app_to_stop_.begin(), navi_app_to_stop_.end(), app_id)) {
```
Perhaps this check should be moved into the `EndNaviServices` function

LitvinenkoIra commented:
@jacobkeeler Thank you for the review.
I will make the necessary changes, check the regression and let you know when this PR is ready for re-review.

jacobkeeler commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -3894,7 +3920,10 @@ void ApplicationManagerImpl::CloseNaviApp() {
+  {
+    sync_primitives::AutoLock lock(navi_app_to_stop_lock_);
+    navi_app_to_stop_.pop_front();
+  }
```
I think this should stay at the top of the function, otherwise the timer could theoretically fire for 2 different apps, but the same app ID would be read twice

jacobkeeler replied:
Ah, I see that this would cause issues, perhaps you could put this value to the end of the queue and then remove it from there after?

LitvinenkoIra replied:
This works, but the changes are not obvious, since this is a regression after https://github.com/smartdevicelink/sdl_core/pull/2425 I will try to rework that fix, and do not call EndService inside the UnregisterApp. I'll let you know when the changes are ready

LitvinenkoIra replied:
@jacobkeeler I forced the commits to make the good commit history. I reworked the fix from #2425 in 898b824f78954871c0a31f5f6dea7f095827f959 and now SDL works as expected in accordance with https://github.com/smartdevicelink/sdl_core/issues/1409 and https://github.com/smartdevicelink/sdl_core/issues/3328
Now I'm checking the regression and I will let you know when it will be passed.

jacobkeeler commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -3863,9 +3891,13 @@ void ApplicationManagerImpl::CloseNaviApp() {
+  {
+    sync_primitives::AutoLock lock(navi_app_to_stop_lock_);
+    DCHECK_OR_RETURN_VOID(!navi_app_to_stop_.empty());
+    app_id = navi_app_to_stop_.front();
```
```suggestion
    app_id = navi_app_to_stop_.front();
    navi_app_to_stop_.pop_front();
    navi_app_to_stop_.push_back(app_id);
```
As a suggestion

jacobkeeler commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -3887,6 +3919,11 @@ void ApplicationManagerImpl::CloseNaviApp() {
+
+  {
+    sync_primitives::AutoLock lock(navi_app_to_stop_lock_);
+    navi_app_to_stop_.pop_front();
```
```suggestion
    auto it = std::find(navi_app_to_stop_.begin(), navi_app_to_stop_.end(), app_id);
    if (navi_app_to_stop_.end() != it) {
        navi_app_to_stop_.erase(it);
    }
```
As a suggestion

jacobkeeler commented:
@LitvinenkoIra We would appreciate if developers did not force push to a PR that is in review. Force pushing makes it difficult to review a PR because we cannot tell what was changed since the last commit. In general, please do not rebase and do not force push against the SDL repositories. Instead, you can merge the current develop branch into your PR branch. Thank you

LitvinenkoIra commented:
@jacobkeeler Ok, got it. Sorry for that, I just removed unnecessary changes, because we don't need all that logic with the `navi_to_stop_` queue. 
I moved the function `EndNaviServices` from `UnregisterApplication` to `UnregisterAllApplications()` and now SDL works as expected. 
You can find the protection for `navi_to_stop_` here 20109cf 
And fix for the issue here 898b824

jacobkeeler commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -3215,8 +3223,6 @@ void ApplicationManagerImpl::UnregisterApplication(
     MessageHelper::SendUnsubscribedWayPoints(*this);
   }
 
-  EndNaviServices(app_id);
```
Now `EndService` will not be sent for other unregister reasons, such as changing the system language
